src/Neo/Network/P2P/Payloads/NotaryAssisted.cs

```
@@ -48,14 +47,19 @@ protected override void SerializeWithoutType(BinaryWriter writer)
 
         public override JObject ToJson()
         {
-            JObject json = base.ToJson();
+            var json = base.ToJson();
             json["nkeys"] = NKeys;
             return json;
         }
 
         public override bool Verify(DataCache snapshot, Transaction tx)
         {
-            return tx.Signers.Any(p => p.Account.Equals(notaryHash));
+            if (tx.Sender == s_notaryHash)
+            {
+                // Payer is in the second position
+                return tx.Signers.Length == 2;
```

It's not needed, Notary contract contains this check in `verify` method:
https://github.com/neo-project/neo/blob/46622dbb320a5d969e4690a194c6b829a244d331/src/Neo/SmartContract/Native/Notary.cs#L120

---

This is faster, we don't use a vm to check it, we can have both

---

> This is faster

Agree, attributes check goes prior to witness check. But then there's a duplication in functionality, we always need to keep this code in sync with Notary's `verify` method.

To me, it doesn't hurt to add this check. @roman-khimov?

---

I'd prefer to avoid duplication, it's just cleaner code-wise. Then:
 * for a good transaction this doesn't make any difference, we have to check both attribute and witness anyway
 * for an unintentionally bad transaction this can fail faster, but that's a rare event usually at development stage, doesn't affect real performance
 * for an intentionally bad transaction this doesn't make any difference since it'll try to suck as much resources as it can and it has more chances for doing that by leveraging invocation/verification scripts rather than poking at notary verification

---

Then it's better here, because it's faster and it's produced before execution avoiding extra costs

---

Quite the contrary to me, it's notary contract that holds the GAS, it's notary contract that must ensure it's used safely for fee payments. Attribute is for notary economics handling mostly, it's not a security mechanism. Imagine we're changing notary contract to handle more than two signers for fallbacks (which technically is quite possible and can be a valid enhancement) --- it shouldn't require any changes to attribute.

Performance-wise see above, there is no practical difference.

---

For me it's different, there are two verifications layers, both are compatible and make it twice doesn't hurt to anyone. It's the attribute valid without this configuration, no, so why we need to execute the vm if it's wrong defined? It's a waste of time and resources.

Also, if you read the code of the attribute, you quickly can understand how is designed to be used, but if you pretend to configure the attribute after reading the contract you wil need a deeper knowledge.

What's your opinion @Jim8y @cschuchardt88 @Wi1l-B0t ?

---

Another option is leave this verification in c# and not in neo-go, I'm sure that not all the verifications are the same in other places

---

> make it twice

is exactly that:

> It's a waste of time and resources

Plus associated code maintenance overhead.

> why we need to execute the vm if it's wrong defined

In reality you have to do that anyway, there is no benefit. Any real good transaction requires that. And if anyone wants to put some stress onto your node by sending invalid transactions he'll certainly make them invoke VM, it's too trivial to do that.

> I'm sure that not all the verifications are the same in other places

All of the security and protocol-level things are the same, otherwise we wouldn't be compatible.

But other opinions are welcome of course. My point is that this double-check adds an unnecessary overhead degrading performance, maintainability and clarity.

---

For me i dont really think it matters at all, but better to keep it the same for both C# and neogo, cause the final gas fee paid might be different....

As to `Imagine we're changing notary contract to handle more than two signers for fallbacks`, we will need to keep in mind to update the logic here everytime we want to update the notrary contract.

---

Maybe just keep it as it is.